### ISSUE TITLE
cpu exception in dosemu code outside of VM86 when launched from .desktop file

### ISSUE THREAD

**Issue by severach** [stsp/dosemu2#414]
````
ERROR: Fault in dosemu code, in_dpmi=0
ERROR: cpu exception in dosemu code outside of VM86()!
````

`dosemu` works when launched from an xterm or a .desktop file with terminal enabled. From .desktop file with terminal disabled `dosemu` crashes. Happens in Cinnamon or XFCE. Old versions of `dosemu` are crashing too so it might have something to do with the new `-fno-plt` option enabled system wide in Arch.

It no longer compiles with clang.
````
clang -c -MMD -DCFLAGS_STR="-Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -Wno-unused-result -fpie -fno-stack-protector -Wno-microsoft -Wno-incompatible-pointer-types -ggdb3   -march=x86-64 -mtune=generic -O2 -pipe  -fno-plt -fno-stack-protector -Wno-unused-result" -I../../../src/include -I../../../src/plugin/include -I/home/chris/build/dosemu2-git/src/dosemu2/src/include -D_FORTIFY_SOURCE=2 -Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -Wno-unused-result -fpie -fno-stack-protector -Wno-microsoft -Wno-incompatible-pointer-types -ggdb3   -march=x86-64 -mtune=generic -O2 -pipe  -fno-plt -fno-stack-protector -Wno-unused-result -o int.o int.c
int.c:2452:1: error: expected identifier or '('
INT_WRP(16)
^
int.c:2440:20: note: expanded from macro 'INT_WRP'
static int _int##n(int stk_offs) \
                   ^
int.c:2452:1: error: expected ')'
int.c:2440:20: note: expanded from macro 'INT_WRP'
static int _int##n(int stk_offs) \
                   ^
int.c:2452:1: note: to match this '('
int.c:2440:19: note: expanded from macro 'INT_WRP'
static int _int##n(int stk_offs) \
                  ^
int.c:2499:54: error: expected expression
  int_handlers[0x16].interrupt_function[NO_REVECT] = _int16;
                                                     ^
3 errors generated.
make[2]: *** [/tmp/makepkg/dosemu2-git/src/dosemu2/src/Makefile.common:62: int.o] Error 1
make[2]: Leaving directory '/tmp/makepkg/dosemu2-git/src/dosemu2/src/base/async'
````

The log also has some wrong information.

````
Your fluidsynth is too old
````
No it's not. I installed enough stuff to make this error go away on launch but it still shows in the log.

````
Reading symbols from /usr/bin/dosemu.bin...(no debugging symbols found)...done.
````
Yes there are. I view with `mc` and I can see all the symbols.

[boot.log.zip](https://github.com/stsp/dosemu2/files/1178425/boot.log.zip)
[boot.log.nocrash.zip](https://github.com/stsp/dosemu2/files/1178428/boot.log.nocrash.zip)

````
% pacman -Q dosemu2-git gcc-multilib glibc sdl2
dosemu2-git 2.0pre6.1.dev.671.g19a8aafd-1
gcc-multilib 7.1.1-4
glibc 2.25-7
sdl2 2.0.5-4
````

**Comment by stsp** [comment]
Grr, clang defines something with the name `_int16`.
Should now be fixed.

> No it's not.

Its yes, see the date of the fluidsynth sources
in your package. The new sources are in their
git, but they never did a release since 2013.
So this message will likely always appear.

> Yes there are. I view with mc and I can see all the symbols.

But this does not mean a debug symbols.
gdb explicitly tells you about the _debugging_
symbols, not about some symbols you can
see in mc. Please trust gdb, if it says so, you
should build with `--enable-debug`. Note that
both rpm and deb builds now use `--enable-debug`,
so I suggest you to do the same.
Since your log does not have debug info, it is
not possible to see where the crash happens.

**Comment by severach** [comment]
More logs
[boot.log.clang.zip](https://github.com/stsp/dosemu2/files/1180695/boot.log.clang.zip)
[boot.log.gcc.zip](https://github.com/stsp/dosemu2/files/1180696/boot.log.gcc.zip)

**Comment by stsp** [comment]
Feel free to open new ticket.
Does it work with `-X`?

**Comment by severach** [comment]
Adding -X to the desktop file makes it launch.

**Comment by stsp** [comment]
Please re-do the log with current devel.

**Comment by severach** [comment]
[boot.log.zip](https://github.com/stsp/dosemu2/files/1181664/boot.log.zip)

**Comment by stsp** [comment]
And now?

**Comment by severach** [comment]
[boot.log.zip](https://github.com/stsp/dosemu2/files/1181738/boot.log.zip)

Are these of any interest to you?

````
signal.c:56:10: warning: #warning Not disabling SIGALTSTACK_WA, update your kernel [-Wcpp]
         #warning Not disabling SIGALTSTACK_WA, update your kernel
          ^~~~~~~
signal.c:91:10: warning: #warning Not disabling SIGRETURN_WA, update your kernel [-Wcpp]
         #warning Not disabling SIGRETURN_WA, update your kernel
          ^~~~~~~
pci.c: In function ‘pci_open_proc’:
pci.c:292:46: warning: ‘%d’ directive writing between 1 and 3 bytes into a region of size 2 [-Wformat-overflow=]
   sprintf(proc_pci_name_buf + 14, "%02x/%02x.%d", bus, device, fn);
                                              ^~
pci.c:292:35: note: directive argument in the range [0, 255]
   sprintf(proc_pci_name_buf + 14, "%02x/%02x.%d", bus, device, fn);
                                   ^~~~~~~~~~~~~~
In file included from /usr/include/stdio.h:939:0,
                 from ../../../../src/include/emu.h:11,
                 from pci.c:16:
/usr/include/bits/stdio2.h:33:10: note: ‘__builtin___sprintf_chk’ output between 8 and 10 bytes into a destination of size 8
   return __builtin___sprintf_chk (__s, __USE_FORTIFY_LEVEL - 1,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       __bos (__s), __fmt, __va_arg_pack ());
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

---------------------------------DONE compiling-------------------------------
````

**Comment by stsp** [comment]
Could you please attach your `/usr/lib/dosemu/libplugin_X.so`?

**Comment by severach** [comment]
This is before the patch.
[libplugin_X.so.zip](https://github.com/stsp/dosemu2/files/1181800/libplugin_X.so.zip)

**Comment by stsp** [comment]
Is the clang warning fixed?

**Comment by stsp** [comment]
Does it work now?

**Comment by stsp** [comment]
`libplugin_X.so` looks fine, all symbols
are exported, not sure why `dlsym()` fails
for you. Can you give me an ssh to the
test machine?

**Comment by severach** [comment]
clang compiles fine but dosemu doesn't work any better. There are less warnings. 1 of 4 computers can launch dosemu. 3 of 4 still can't.

dosemu launches without crashing from a terminal. You need a desktop and a clicky start menu to crash it. I can supply an ssh but I don't see how it would help. You can install and update Manjaro in about an hour then install dosemu2-git.

**Comment by stsp** [comment]
Can you try:
`valgrind --trace-children=yes dosemu`
Maybe valgrind will find something interesting...

**Comment by stsp** [comment]
Also please do the following:
```
diff --git a/src/plugin/sdl/keyb_SDL.c b/src/plugin/sdl/keyb_SDL.c
index c22035d..e003420 100644
--- a/src/plugin/sdl/keyb_SDL.c
+++ b/src/plugin/sdl/keyb_SDL.c
@@ -36,6 +36,7 @@
 #ifndef USE_DL_PLUGINS
 #undef X_SUPPORT
 #endif
+#define HAVE_XKB 0
 
 /*
  * at startup SDL does not detect numlock and capslock status
```

and produce the log with this change please.

**Comment by severach** [comment]
With patch [boot.log.zip](https://github.com/stsp/dosemu2/files/1184542/boot.log.zip)

valgrind doesn't work in the desktop file. That would debug all the start up which isn't a problem. Can you patch /usr/bin/dosemu with a commented line that runs valgrind?

**Comment by stsp** [comment]
This log does not seem to have any crash info.

**Comment by severach** [comment]
Thats with the patch. I added `exec 2> ~/dosemu.log` to the top of `/usr/bin/dosemu`. It doesn't crash or launch a window. It doesn't do anything unless I add `-X` to the .desktop file.

**Comment by stsp** [comment]
Yes, I can patch dosemu script:
```
diff --git a/src/bindist/dosemu b/src/bindist/dosemu
index f279894..a133c5e 100644
--- a/src/bindist/dosemu
+++ b/src/bindist/dosemu
@@ -154,6 +154,8 @@ if [ -n "$STRING_I" ] ; then
   COMMAND="$COMMAND -I $STRING_I"
 fi
 
+COMMAND="valgrind --trace-children=yes $COMMAND"
+
 exec 3>&1
 # in a non-bash case, $BASHPID won't work and we fail the cleanup - no problem
 DPID=$(echo $BASHPID; exec 1<&3; exec $COMMAND "$@")
```

By the way, is it a glibc system, or musl?

**Comment by stsp** [comment]
For valgrind to work properly, you need
the following:
```
$_cpu_emu = "vm86sim"
$_cpu_vm = "emulated"
```

But to check for a bug, maybe you don't
even need this.

**Comment by severach** [comment]
Without those config lines I had to ^C^C to get out. I added `exec 2> ~/.dosemu/dosemu.stderr.log`.

[boot.log.valgrind.zip](https://github.com/stsp/dosemu2/files/1186076/boot.log.valgrind.zip)

**Comment by stsp** [comment]
@jschwartzenberg maybe you have some
time for reproducing this?

**Comment by stsp** [comment]
@severach could you please add the new crash log?

**Comment by severach** [comment]
[boot.log-2017-08-22.zip](https://github.com/stsp/dosemu2/files/1243167/boot.log-2017-08-22.zip)
[boot.log-2017-08-22b.zip](https://github.com/stsp/dosemu2/files/1243176/boot.log-2017-08-22b.zip)

**Comment by stsp** [comment]
Please use `-D9+cv`

**Comment by stsp** [comment]
Could you please do the following:
```
$ gdb dosemu.bin
(gdb) b SDL_priv_init
Function "SDL_priv_init" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (SDL_priv_init) pending.
(gdb) b sdl_kbd_init
Function "sdl_kbd_init" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 2 (sdl_kbd_init) pending.
(gdb) r
...
Breakpoint 1, SDL_priv_init () at sdl.c:208
208	{
(gdb) wat X_handle
Hardware watchpoint 3: X_handle
(gdb) c
Continuing.
```

and whenever it stops and prints something, press
`c` and ENTER. Then copy the results here.

**Comment by severach** [comment]
[boot.log-2017-08-22c.zip](https://github.com/stsp/dosemu2/files/1243773/boot.log-2017-08-22c.zip)

dosemu doesn't crash if launched from a terminal or with the terminal enabled in the desktop file so I can't control anything interactive. I tried to make it crash with lines like `bash -c 'exec 0>&- 2>&-; dosemu.bin'` but dosemu won't crash. An exec 1> or 2> at the top of dosemu is the most I've been able to do and keep it crashing. It could work if there was a way to automate the Q&A or attach gdb to another terminal.
